File: medium/47.permutations-ii.java
import java.util.*;

/*
 * @lc app=leetcode id=47 lang=java
 *
 * [47] Permutations II
 */

// @lc code=start
class Solution {
    private List<List<Integer>> ans = new ArrayList<List<Integer>>();
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        Deque<Integer> path = new LinkedList<>();
        boolean[] used = new boolean[nums.length];
        backTracking(nums, used, path);
        return ans;
    }

    private void backTracking(int[] nums, boolean[] used, Deque<Integer> path) {
        // exit
        if (path.size() == nums.length) {
            ans.add(new ArrayList<Integer>(path));
            return;
        }

        for (int i=0; i<nums.length; i++) {
            if (i > 0 && nums[i] == nums[i-1] && used[i-1]) {
                continue;
            }

            if (used[i]) {
                continue;
            }

            used[i] = true;
            path.push(nums[i]);

            backTracking(nums, used, path);

            path.pop();
            used[i] = false;
        }
    }
}
// @lc code=end

File: medium/47.permutations-ii.py
#
# @lc app=leetcode id=47 lang=python3
#
# [47] Permutations II
#

# @lc code=start
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def dfs(nums: List[int], path: List[int], visited: Set[int]):
            if len(path) == len(nums):
                ans.append(path.copy())
                return
            
            for i, num in enumerate(nums):
                if i > 0 and nums[i] == nums[i-1] and i-1 in visited:
                    continue

                if i in visited:
                    continue

                path.append(num)
                visited.add(i)
                dfs(nums, path, visited)
                visited.remove(i)
                path.pop()
            

        ans = []
        path = []
        visited = set()
        nums.sort()
        dfs(nums, path, visited)

        return ans

        
# @lc code=end

